src/functions/org/apache/jmeter/functions/TimeShift.java

```diff
@@ -0,0 +1,188 @@
+ * limitations under the License.
+ *
+ */
+
+package org.apache.jmeter.functions;
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.Year;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeFormatterBuilder;
+import java.time.format.DateTimeParseException;
+import java.time.temporal.ChronoField;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.jmeter.engine.util.CompoundVariable;
+import org.apache.jmeter.samplers.SampleResult;
+import org.apache.jmeter.samplers.Sampler;
+import org.apache.jmeter.threads.JMeterVariables;
+import org.apache.jmeter.util.JMeterUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * timeShifting Function permit to shift a date
+ *
+ * Parameters: 
+ * - format date @see
+ * https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html
+ * (optional - defaults to epoch time in millisecond) 
+ * - date to shift formated
+ * as first param (optional - defaults now) 
+ * - amount of (seconds, minutes,
+ * hours, days or months ) to add (optional - default nothing is add ) 
+ * - variable name ( optional )
+ *
+ * Returns: a formatted date with the specified number of (seconds,
+ * minutes, hours, days or months ) added. - value is also saved in the
+ * variable for later re-use.
+ *
+ * @since 3.3
+ */
+public class TimeShift extends AbstractFunction {
+    private static final Logger log = LoggerFactory.getLogger(TimeShift.class);
+
+    private static final String KEY = "__timeShift"; // $NON-NLS-1$
+
+    private static final List<String> desc = Arrays.asList(
+            JMeterUtils.getResString("time_format_shift"),
+            JMeterUtils.getResString("date_to_shift"),
+            JMeterUtils.getResString("value_to_shift"),
+            JMeterUtils.getResString("function_name_paropt"));
+
+    // Ensure that these are set, even if no paramters are provided
+    private String format = ""; //$NON-NLS-1$
+    private String dateToShift = ""; //$NON-NLS-1$
+    private String amountToShift = ""; //$NON-NLS-1$
+    private String variableName = ""; //$NON-NLS-1$
+    private Pattern pattern = Pattern.compile("(?:((?:[+-])?\\d+)([MdHms]))");
+    
+    public TimeShift() {
+        super();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String execute(SampleResult previousResult, Sampler currentSampler) throws InvalidVariableException {
+        String dateString;
+        LocalDateTime localDateTimeToShift = LocalDateTime.now(ZoneId.systemDefault());
+        DateTimeFormatter formatter = null;
+        if (!StringUtils.isEmpty(format)) {
+            try {
+                formatter = new DateTimeFormatterBuilder().appendPattern(format)
```

Comment:
I had a caffeine cache but I don't know how to set it ( size / expiration ... ? )